compiler/rustc_errors/src/lib.rs

```diff
@@ -993,25 +994,25 @@ impl HandlerInner {
         self.taught_diagnostics.insert(code.clone())
     }
 
-    fn force_print_diagnostic(&mut self, db: Diagnostic) {
-        self.emitter.emit_diagnostic(&db);
+    fn force_print_diagnostic(&mut self, mut db: Diagnostic) {
+        self.emitter.emit_diagnostic(&mut db);
     }
 
     /// Emit all stashed diagnostics.
     fn emit_stashed_diagnostics(&mut self) -> Option<ErrorGuaranteed> {
         let diags = self.stashed_diagnostics.drain(..).map(|x| x.1).collect::<Vec<_>>();
         let mut reported = None;
-        diags.iter().for_each(|diag| {
+        for mut diag in diags {
             if diag.is_error() {
                 reported = Some(ErrorGuaranteed(()));
             }
-            self.emit_diagnostic(diag);
-        });
+            self.emit_diagnostic(&mut diag);
+        }
         reported
     }
 
     // FIXME(eddyb) this should ideally take `diagnostic` by value.
-    fn emit_diagnostic(&mut self, diagnostic: &Diagnostic) -> Option<ErrorGuaranteed> {
+    fn emit_diagnostic(&mut self, diagnostic: &mut Diagnostic) -> Option<ErrorGuaranteed> {
```

Comment:
Would fixing the `FIXME` be possible as part of this PR, or are there big consequences that I'm failing to think of?

Reply:
This requires to make the `DiagnosticBuilder::emit` method take `self` by move, which breaks `struct_err(..).label(..).emit()` chains.

Reply:
Not exactly, I decided we can't make `.emit()` take `&mut self` for that reason, but once `.emit()` is called the `DiagnosticBuilder` state changes to ignore the `Diagnostic` on further `.emit()`, so we could be replacing the `Diagnostic` with a dummy one (or even having `Option` around it but that could slow down the `Deref`/`DerefMut` impl a bunch).